Improve http error codes

veracioux opened the pull request:
This PR refactors error handling in the Opencode server and storage modules to use more appropriate HTTP status.

Key changes:

- ~Adds HTTP status code support to NamedError and~ updates error responses to use correct codes (400, 404, 500).
- 500 (Internal Server Error) is now used instead of 400 (Bad Request) for unknown errors
- Implements NotFoundError in storage operations, returning 404 when resources are missing.
- ~Adds httpCode field to NamedError, so the proper HTTP status code can be assigned to each error at time of definition (falling back to 500)~
- Updates OpenAPI spec to document new error responses.
- Ensures ModelNotFoundError uses status 400.

Example payload:
```json
{
  "name": "NotFoundError",
  "data": {
    "message": "Resource not found: /home/haris/.local/share/opencode/storage/session/ea2235cd982d8eaa572f68a5f2537637b1d1a2c2/ses_a.json"
  }
}
```

Closes #3135.

veracioux committed:
refactor: Add HTTP status code support to NamedError

veracioux committed:
feat: Add NotFoundError handling to storage operations

- Create NotFoundError class with 404 status code
- Wrap storage operations (read, write, update, remove) in error handler

veracioux committed:
fix: Use appropriate HTTP status codes for error responses

- Use err.httpCode when available instead of hardcoded 400
- Return 500 for internal server errors instead of 400

veracioux committed:
docs: Add appropriate errors to OpenAPI spec

- Replace blanket ERRORS object usage with selective error() helper
- Add 404 NotFoundError schema for storage-related endpoints
- Refine 400 BadRequestError schema to include errors array and success flag
- Most endpoints now document both 400 and 404 responses where applicable

veracioux committed:
fix: Set ModelNotFoundError HTTP status to 400

rekram1-node commented on packages/opencode/src/util/error.ts:
```diff
@@ -6,8 +7,9 @@ import z from "zod/v4"
+  abstract httpCode: ContentfulStatusCode
 
-  static create<Name extends string, Data extends z.core.$ZodType>(name: Name, data: Data) {
+  static create<Name extends string, Data extends z.core.$ZodType>(name: Name, data: Data, httpCode: ContentfulStatusCode = 500) {
```
The thing I don't love here is that it can take away from some ddd, currently the server is super important ofc but not every client is over http, if I use `opencode run` and use the --session <id> I wouldn't expect to see a 404 status code in the terminal necessarily? 

I think the http status codes should be handled at http server and all the errors should be agnostic of that

veracioux replied:
You shouldn't see it in the terminal anyway - the httpCode is only used when deciding a HTTP response status code. That's why I deliberately didn't put httpCode in the schema of the error.

However, I do get your point, and there aren't many error classes for which the error code can be known upfront, so I don't feel strongly about it either way. I'll make a few updates.

veracioux replied:
Done

veracioux committed:
refactor: Move HTTP status code logic to onError handler

- Remove httpCode parameter from NamedError.create()
- Remove hardcoded HTTP codes from error definitions
  (Storage.NotFoundError, Provider.ModelNotFoundError)
- Centralize HTTP status code mapping in onError handler

rekram1-node commented:
amazing thanks!
